File: python_system/NetworkReceiver.py
import socket
from dotenv import load_dotenv
import os

load_dotenv()

CLIENT_IN_PORT = int(os.getenv('CLIENT_IN_PORT'))


class NetworkReceiver():
    """
    Class to receive data from a UDP socket.
    """
    def __init__(self, host='localhost', port=CLIENT_IN_PORT):
        self.host = host
        self.port = port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.bind((self.host, self.port))

    def receive_data(self):
        data, addr = self.sock.recvfrom(1024)  # buffer size is 1024 bytes
        return data.decode('utf-8')

File: hft_system/src/NetworkReceiver.cpp
#include "NetworkReceiver.h"

#include <arpa/inet.h>
#include <unistd.h>
#include <iostream>
#include <cstring>
#include <stdexcept>

#define BUFFER_SIZE 1024

NetworkReceiver::NetworkReceiver(int port) {
    sockfd = socket(AF_INET, SOCK_DGRAM, 0);

    if (sockfd < 0) {
        perror("socket creation failed");
        throw std::runtime_error("Failed to create socket");
    }

    memset(&servaddr, 0, sizeof(servaddr));
    memset(&cliaddr, 0, sizeof(cliaddr));

    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = INADDR_ANY;
    servaddr.sin_port = htons(port);

    // Bind the socket
    if (bind(sockfd, (const struct sockaddr *)&servaddr, sizeof(servaddr)) < 0) {
        perror("bind failed");
        close(sockfd);
        throw std::runtime_error("Failed to bind socket");
    }

    std::cout << "Receiver listening on port " << port << "...\n";
}

NetworkReceiver::~NetworkReceiver() {
    close(sockfd);
}

std::string NetworkReceiver::receive_data() {
    char buffer[BUFFER_SIZE];
    socklen_t len = sizeof(cliaddr);

    int n = recvfrom(sockfd, buffer, BUFFER_SIZE - 1, 0,
                     (struct sockaddr *)&cliaddr, &len);
    if (n < 0) {
        perror("recvfrom failed");
        throw std::runtime_error("Error receiving data");
    }

    buffer[n] = '\0';  // Null-terminate
    return std::string(buffer);
}


